felix/calc/encapsulation_resolver.go

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2022 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package calc
+
+import (
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/felix/config"
+	"github.com/projectcalico/calico/felix/dispatcher"
+
+	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+
+	"github.com/projectcalico/calico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/calico/libcalico-go/lib/backend/encap"
+	"github.com/projectcalico/calico/libcalico-go/lib/backend/model"
+)
+
+// EncapsulationResolver is a Calculation Graph component that watches IP pool updates and
+// calculates if the IPIP or VXLAN encaps should be enabled or disabled. If there is
+// a change, Felix restarts by calling configChangedRestartCallback() to apply them.
+type EncapsulationResolver struct {
+	config                       *config.Config
+	encapCalc                    *EncapsulationCalculator
+	configChangedRestartCallback func()
+	inSync                       bool
+}
+
+func NewEncapsulationResolver(config *config.Config, configChangedRestartCallback func()) *EncapsulationResolver {
+	if configChangedRestartCallback == nil {
+		log.Panic("Starting EncapsulationResolver with nil callback func.")
+	}
+
+	return &EncapsulationResolver{
+		config:                       config,
+		encapCalc:                    NewEncapsulationCalculator(config, nil),
+		configChangedRestartCallback: configChangedRestartCallback,
+		inSync:                       false,
+	}
+}
+
+func (r *EncapsulationResolver) RegisterWith(dispatcher *dispatcher.Dispatcher) {
+	dispatcher.Register(model.IPPoolKey{}, r.OnPoolUpdate)
+	dispatcher.RegisterStatusHandler(r.OnStatusUpdate)
+}
+
+func (r *EncapsulationResolver) OnPoolUpdate(update api.Update) (filterOut bool) {
+	if update.Value == nil {
+		log.WithField("update", update).Debug("EncapsulationResolver: IPPool deletion")
+
+		k, ok := update.Key.(model.IPPoolKey)
+		if ok {
+			r.encapCalc.RemoveModelPool(k)
+		} else {
+			log.Infof("failed to convert %+v to model.IPPoolKey. Ignoring.", update.Key)
```

Comment:
We'll call TriggerCalculation even if we hit this branch - is that desirable? Probably not a big issue, but is extra work that's not needed technically.

Reply:
I would probably structure this check like this to be a little bit less verbose and also more clear that there's no work to do on invalid updates. 


```go
k, ok := update.Key.(model.IPPoolKey)
if !ok {
  log.Error("failed to convert object to model.IPPoolKey, ignoring")
  return
}
r.encapCalc.RemoveModelPool(k)

Reply:
good point, thanks! will do!